### ISSUE TITLE
Non-requested "tools" JSON object sent

### ISSUE THREAD

**Issue by UnaDeKalamares** [SimonSchubert/Kai#283]
### Description

I've connected Kai 9000 to LM Studio via Open-API. Regardless of the length of the message, it always sends an unnecessary long message with non-requested content in the JSON payload. This results in the call being rejected, as it's exceeding the current context.

### Steps to Reproduce

1) Connect to LM Studio via Open-API endpoint
2) Try sending "hello?"
3) Call rejected.

### Kai Version

2.6.2

### Platform

Android

### Logs / Screenshots

This is the output log from LM Studio:

```
2026-05-21 19:04:46 [DEBUG]
 Received request: POST to /v1/chat/completions with body  {
  "messages": [
    {
      "role": "system",
      "content": "You're not a chatbot. You're a personal assistant ... <Truncated in logs> ...\",\"event\":\"submit\",\"collectFrom\":[\"name\"]}}]}\n```\n"
    },
    {
      "role": "user",
      "content": "hello?"
    }
  ],
  "model": "qwen/qwen3-coder-30b",
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "execute_shell_command",
        "description": "Execute a shell command in an Alpine Linux sandbox... <Truncated in logs> ...ath=\"page.html\"). File needs to be self-contained.",
        "parameters": {
          "type": "object",
          "properties": {
            "command": {
              "type": "string",
              "description": "The shell command to execute"
            },
            "timeout": {
              "type": "integer",
              "description": "Timeout in seconds (default 30, max 60)"
            },
            "working_dir": {
              "type": "string",
              "description": "If set, run the command starting in this directory... <Truncated in logs> ...calls — same as if the user had run cd themselves."
            },
            "env": {
              "type": "object",
              "description": "Per-command environment variable overrides. Scoped... <Truncated in logs> ...port' inside the command if you want persistence)."
            },
            "background": {
              "type": "boolean",
              "description": "Run detached as a background job. Returns a sessio... <Truncated in logs> ...check status. Does not share the persistent shell."
            },
            "fresh": {
              "type": "boolean",
              "description": "If true, run in a one-shot isolated shell that does not share state with the persistent session. Default false."
            }
          },
          "required": [
            "command"
          ]
        }
      }
    },
    {
      "type": "function",
      "function": {
        "name": "manage_process",
        "description": "Manage background shell processes started with exe... <Truncated in logs> ...inished process from the list (params: session_id)",
        "parameters": {
          "type": "object",
          "properties": {
            "action": {
              "type": "string",
              "description": "Action to perform: list, log, kill, or remove"
            },
            "session_id": {
              "type": "string",
              "description": "Session ID of the process (required for log, kill, remove)"
            },
            "offset": {
              "type": "integer",
              "description": "Line offset for log output (default: 0)"
            },
            "limit": {
              "type": "integer",
              "description": "Max lines to return for log (default: 200)"
            }
          },
          "required": [
            "action"
          ]
        }
      }
    },
    {
      "type": "function",
      "function": {
        "name": "ssh_configure_host",
        "description": "Register a named SSH host alias in the Linux sandb... <Truncated in logs> ...alias bypasses every setting this tool just wrote.",
        "parameters": {
          "type": "object",
          "properties": {
            "alias": {
              "type": "string",
              "description": "Short name used to invoke this host (e.g. 'prod', 'my-vps'). Must contain no whitespace."
            },
            "hostname": {
              "type": "string",
              "description": "DNS name or IP of the remote machine."
            },
            "user": {
              "type": "string",
              "description": "SSH user. Omit to fall back to the openssh default (current user inside the sandbox)."
            },
            "port": {
              "type": "integer",
              "description": "SSH port. Omit for 22."
            },
            "identity_file": {
              "type": "string",
              "description": "Path to the private key inside the sandbox. Relati... <Truncated in logs> ...ass through. The file is not created by this tool."
            },
            "known_host_line": {
              "type": "string",
              "description": "Optional single line appended to ~/.ssh/known_host... <Truncated in logs> ...ed25519 <host>`. Deduplicated by exact-line match."
            }
          },
          "required": [
            "alias",
            "hostname"
          ]
        }
      }
    }
  ]
}
2026-05-21 19:04:46  [INFO]
 [qwen/qwen3-coder-30b] Running chat completion on conversation with 2 messages.
2026-05-21 19:04:46 [DEBUG]
 LlamaV4::predict slot selection: session_id=<empty> server-selected (LCP/LRU)
2026-05-21 19:04:46 [DEBUG]
 slot get_availabl: id  1 | task -1 | selected slot by LRU, t_last = 30699854048
srv  get_availabl: updating prompt cache
srv          load:  - looking for better prompt, base f_keep = -1.000, sim = 0.000
srv        update:  - cache state: 6 prompts, 82.986 MiB (limits: 8192.000 MiB, 4096 tokens, 87363 est)
srv        update:    - prompt 0x7f8e2c314810:     149 tokens, checkpoints:  0,    13.972 MiB
srv        update:    - prompt 0x7f8e2c32d150:      33 tokens, checkpoints:  0,     3.095 MiB
srv        update:    - prompt 0x7f8e1c438070:      76 tokens, checkpoints:  0,     7.127 MiB
srv        update:    - prompt 0x7f8e2c008230:     110 tokens, checkpoints:  0,    10.315 MiB
srv        update:    - prompt 0x7f8e2c493b80:      92 tokens, checkpoints:  0,     8.627 MiB
srv        update:    - prompt 0x7f8e2c27c2e0:     425 tokens, checkpoints:  0,    39.850 MiB
srv  get_availabl: prompt cache update took 0.01 ms
2026-05-21 19:04:46 [DEBUG]
 slot launch_slot_: id  1 | task -1 | sampler chain: logits -> penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> ?xtc -> temp-ext -> dist 
slot launch_slot_: id  1 | task 501 | processing task, is_child = 0
slot update_slots: id  1 | task 501 | new prompt, n_ctx_slot = 4096, n_keep = 4848, task.n_tokens = 4848
srv    send_error: task id = 501, error: request (4848 tokens) exceeds the available context size (4096 tokens), try increasing it
slot      release: id  1 | task 501 | stop processing: n_tokens = 0, truncated = 0
srv  update_slots: no tokens to decode
srv  update_slots: all slots are idle
2026-05-21 19:04:46 [ERROR]
 [qwen/qwen3-coder-30b] Error: Channel Error
```

**Comment by TimeJunkie** [comment]
This actually looks like a context size configuration issue on the LM Studio side rather than a Kai bug.

From your logs:

```
n_ctx_slot = 4096
task.n_tokens = 4848
```

Kai's system prompt plus the tool definitions it sends totals ~4848 tokens, which is just over LM Studio's default 4096-token context window.

**To fix it**, increase the context size in LM Studio for that model:

1. Open LM Studio and go to your loaded model's settings
2. Find the **Context Length** setting (sometimes labeled `n_ctx`)
3. Set it to at least **8192** to give yourself comfortable headroom
4. Reload/restart the model and try again

Note that a larger context will use more VRAM, so if you're tight on memory you may need to reduce other settings (like layers offloaded to GPU or experts cpu offloading for better performance for a moe like qwen coder) to compensate.

The main tools you're seeing there are the Alpine Linux tools. So if you have that enabled, they're going to be sending even if you're only saying hello. That's just how it allows the LLM to know what tools are available for it in case you ask for something that could use a tool call. 

---

On another note, @SimonSchubert , I wonder if on Kai's documentation, if it doesn't already, perhaps it should say that the minimum context size recommended for openai compatible servers is 8192 or perhaps 16k even, honestly. I normally am running anywhere from 64K to 128K+ on my server for deep context usage.

**Comment by UnaDeKalamares** [comment]
Thank you @TimeJunkie for your thorough and educational response! I'm an old-fashioned developer trying self-hosted LLMs to learn more about them, and I definitely didn't think of context length when setting this up. I've just tried again, making sure I turn everything off inside the Android app. That would be:

- Left the default "Soul";
- Everything in "Agent" turned off;
- Everything in "Tools" turned off;
- Linux Sandbox not installed;
- Every integration ("Splinterlands") turned off

However, I still notice a big payload, but this time below the default context threshold, so it works as expected:

```
2026-05-22 20:16:11 [DEBUG]
 Received request: POST to /v1/chat/completions with body  {
  "messages": [
    {
      "role": "system",
      "content": "You're not a chatbot. You're a personal assistant ... <Truncated in logs> ...droid\n- Model: \n- Provider: OpenAI-Compatible API\n"
    },
    {
      "role": "user",
      "content": "hello?"
    }
  ],
  "model": "qwen/qwen3-coder-30b",
  "tools": [
    {
      "type": "function",
      "function": {
        "name": "execute_shell_command",
        "description": "Execute a shell command in an Alpine Linux sandbox... <Truncated in logs> ...ath=\"page.html\"). File needs to be self-contained.",
        "parameters": {
          "type": "object",
          "properties": {
            "command": {
              "type": "string",
              "description": "The shell command to execute"
            },
            "timeout": {
              "type": "integer",
              "description": "Timeout in seconds (default 30, max 60)"
            },
            "working_dir": {
              "type": "string",
              "description": "If set, run the command starting in this directory... <Truncated in logs> ...calls — same as if the user had run cd themselves."
            },
            "env": {
              "type": "object",
              "description": "Per-command environment variable overrides. Scoped... <Truncated in logs> ...port' inside the command if you want persistence)."
            },
            "background": {
              "type": "boolean",
              "description": "Run detached as a background job. Returns a sessio... <Truncated in logs> ...check status. Does not share the persistent shell."
            },
            "fresh": {
              "type": "boolean",
              "description": "If true, run in a one-shot isolated shell that does not share state with the persistent session. Default false."
            }
          },
          "required": [
            "command"
          ]
        }
      }
    },
    {
      "type": "function",
      "function": {
        "name": "manage_process",
        "description": "Manage background shell processes started with exe... <Truncated in logs> ...inished process from the list (params: session_id)",
        "parameters": {
          "type": "object",
          "properties": {
            "action": {
              "type": "string",
              "description": "Action to perform: list, log, kill, or remove"
            },
            "session_id": {
              "type": "string",
              "description": "Session ID of the process (required for log, kill, remove)"
            },
            "offset": {
              "type": "integer",
              "description": "Line offset for log output (default: 0)"
            },
            "limit": {
              "type": "integer",
              "description": "Max lines to return for log (default: 200)"
            }
          },
          "required": [
            "action"
          ]
        }
      }
    },
    {
      "type": "function",
      "function": {
        "name": "ssh_configure_host",
        "description": "Register a named SSH host alias in the Linux sandb... <Truncated in logs> ...alias bypasses every setting this tool just wrote.",
        "parameters": {
          "type": "object",
          "properties": {
            "alias": {
              "type": "string",
              "description": "Short name used to invoke this host (e.g. 'prod', 'my-vps'). Must contain no whitespace."
            },
            "hostname": {
              "type": "string",
              "description": "DNS name or IP of the remote machine."
            },
            "user": {
              "type": "string",
              "description": "SSH user. Omit to fall back to the openssh default (current user inside the sandbox)."
            },
            "port": {
              "type": "integer",
              "description": "SSH port. Omit for 22."
            },
            "identity_file": {
              "type": "string",
              "description": "Path to the private key inside the sandbox. Relati... <Truncated in logs> ...ass through. The file is not created by this tool."
            },
            "known_host_line": {
              "type": "string",
              "description": "Optional single line appended to ~/.ssh/known_host... <Truncated in logs> ...ed25519 <host>`. Deduplicated by exact-line match."
            }
          },
          "required": [
            "alias",
            "hostname"
          ]
        }
      }
    }
  ]
}
2026-05-22 20:16:11  [INFO]
 [qwen/qwen3-coder-30b] Running chat completion on conversation with 2 messages.
2026-05-22 20:16:11 [DEBUG]
 LlamaV4::predict slot selection: session_id=<empty> server-selected (LCP/LRU)
2026-05-22 20:16:11 [DEBUG]
 slot get_availabl: id  3 | task -1 | selected slot by LCP similarity, sim_best = 0.202 (> 0.100 thold), f_keep = 0.193
srv  get_availabl: updating prompt cache
2026-05-22 20:16:11 [DEBUG]
 srv   prompt_save:  - saving prompt with length 3274, total state size = 306.976 MiB
2026-05-22 20:16:11 [DEBUG]
 srv          load:  - looking for better prompt, base f_keep = 0.193, sim = 0.202
srv        update:  - cache state: 1 prompts, 306.976 MiB (limits: 8192.000 MiB, 4096 tokens, 87370 est)
srv        update:    - prompt 0x7fd66c00bb60:    3274 tokens, checkpoints:  0,   306.976 MiB
srv  get_availabl: prompt cache update took 121.40 ms
```

 In any case, and correct me if my understanding is incorrect/incomplete, but the query should not contain anything else other than my prompt if nothing else in the app is selected. I think I understand that this is the "LLM way" to define behavior so that the model can understand options inside the payload.

EDIT: changed the displayed payload, since I think I originally pasted something different, apologies!

**Comment by TimeJunkie** [comment]
Ah, if alpine environment is disabled and yet still sending those tool definitions that is a valid concern. 
Sorry @UnaDeKalamares , I guess I misunderstood part of the original issue you were seeing. Certainly increasing the context size like I recommended would still be a good idea, but it is still an issue that it's including extra tools that should not be there possibly - both for context size as well as it could potentially try to call tools that are not even enabled.

@SimonSchubert this does seem like the alpine sandbox tools are not being filtered out from the tool options payload correctly? Perhaps not all the sandbox tools are tagged as part of the toggle for enabling it or not.

**Comment by SimonSchubert** [comment]
@UnaDeKalamares thank you for the detailed information and @TimeJunkie thank you for helping to debug the problem

I was able to reproduce the unexpected tool definitions for turned off linux sandbox. definitely a bug and I'm glad you reported this 👍 

I got this fixed + I noticed the 

```
{
      "role": "system",
      "content": "You're not a chatbot. You're a personal assistant ... <Truncated in logs> ...droid\n- Model: \n- Provider: OpenAI-Compatible API\n"
    },
```

contains memory, tool, task information even with those turned off. fix coming soon

**Comment by SimonSchubert** [comment]
✅ please re-open if issue persists after next app update

https://github.com/SimonSchubert/Kai/commit/1fe80baaeaad40ff622748fb3796765b90d081d2
